presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowStateFactory.java

```diff
@@ -75,28 +80,52 @@ public Slice getLongDecimal()
         public void setLongDecimal(Slice value)
         {
             requireNonNull(value, "value is null");
-            if (getLongDecimal() == null) {
+            if (unscaledDecimals.getAndSet(getGroupId(), value) == null) {
                 numberOfElements++;
             }
-            unscaledDecimals.set(getGroupId(), value);
         }
 
         @Override
         public long getOverflow()
         {
+            if (overflows == null) {
+                return 0;
+            }
             return overflows.get(getGroupId());
         }
 
         @Override
         public void setOverflow(long overflow)
         {
-            overflows.set(getGroupId(), overflow);
+            // setOverflow(0) must overwrite any existing overflow value
+            if (overflow == 0 && overflows == null) {
+                return;
+            }
+            long groupId = getGroupId();
+            if (overflows == null) {
+                overflows = new LongBigArray();
+                overflows.ensureCapacity(unscaledDecimals.getCapacity());
+            }
+            overflows.set(groupId, overflow);
+        }
+
+        @Override
+        public void addOverflow(long overflow)
+        {
+            if (overflow != 0) {
```

Comment:
nit: We could return early for the overflow = 0 case to reduce the indent for the majority of this function.
Reducing the indent could improve the readability of the code.

```
if (overflow == 0) {
  return;
}
long groupId = getGroupId();
if (overflows == null) {
    overflows = new LongBigArray();
    overflows.ensureCapacity(unscaledDecimals.getCapacity());
}
overflows.add(groupId, overflow);
```

Reply:
Done